Your pnpm version is incompatible with "renovate@37.303.2(typanion@3.14.0)".

### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

9.0.3

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI

### Link to the code that reproduces this issue or a replay of the bug

_No response_

### Reproduction steps

- Install `renovate@37.303.2` with npm@8
- Bump pnpm to 9.0.3
- Run `pnpm install`

```
 ERR_PNPM_UNSUPPORTED_ENGINE  Unsupported environment (bad pnpm and/or Node.js version)

Your pnpm version is incompatible with "renovate@37.303.2(typanion@3.14.0)".

Expected version: ^8.0.0
Got: 9.0.3

This is happening because the package's manifest has an engines.pnpm field specified.
To fix this issue, install the required pnpm version globally.

To install the latest version of pnpm, run "pnpm i -g pnpm".
To check your pnpm version, run "pnpm -v".
```

### Describe the Bug

It appears that pnpm cares about the dependencies' manifest has an engines.pnpm field specified.

### Expected Behavior

Pnpm should not care about the engines.pnpm of the dependenices.

### Which Node.js version are you using?

v18.19.0

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

--- a/.changeset/moody-geckos-bathe.md
+++ b/.changeset/moody-geckos-bathe.md
@@ -1,0 +1,9 @@
+---
+"@pnpm/package-requester": patch
+"@pnpm/filter-lockfile": patch
+"@pnpm/headless": patch
+"@pnpm/deps.graph-builder": patch
+"pnpm": patch
+---
+
+The `engines.pnpm` field in the `package.json` files of dependencies should be ignored [#7965](https://github.com/pnpm/pnpm/issues/7965).

--- a/deps/graph-builder/src/lockfileToDepGraph.ts
+++ b/deps/graph-builder/src/lockfileToDepGraph.ts
@@ -120,7 +120,6 @@ export async function lockfileToDepGraph (
             lockfileDir: opts.lockfileDir,
             nodeVersion: opts.nodeVersion,
             optional: pkgSnapshot.optional === true,
-            pnpmVersion: opts.pnpmVersion,
             supportedArchitectures: opts.supportedArchitectures,
           }) === false
         ) {

--- a/lockfile/filter-lockfile/src/filterLockfileByImportersAndEngine.ts
+++ b/lockfile/filter-lockfile/src/filterLockfileByImportersAndEngine.ts
@@ -168,7 +168,6 @@ function pkgAllDeps (
           lockfileDir: opts.lockfileDir,
           nodeVersion: opts.currentEngine.nodeVersion,
           optional: pkgSnapshot.optional === true,
-          pnpmVersion: opts.currentEngine.pnpmVersion,
           supportedArchitectures: opts.supportedArchitectures,
         }) !== false
       if (!installable) {

--- a/pkg-manager/core/test/install/optionalDependencies.ts
+++ b/pkg-manager/core/test/install/optionalDependencies.ts
@@ -161,7 +161,7 @@ test('skip optional dependency that does not support the current Node version', 
   expect(reportedTimes).toBe(1)
 })
 
-test('skip optional dependency that does not support the current pnpm version', async () => {
+test('do not skip optional dependency that does not support the current pnpm version', async () => {
   const project = prepareEmpty()
   const reporter = sinon.spy()
 
@@ -175,7 +175,7 @@ test('skip optional dependency that does not support the current pnpm version', 
     pnpmVersion: '4.0.0',
   }))
 
-  project.hasNot('@pnpm.e2e/for-legacy-pnpm')
+  project.has('@pnpm.e2e/for-legacy-pnpm')
   project.storeHas('@pnpm.e2e/for-legacy-pnpm', '1.0.0')
 
   const logMatcher = sinon.match({
@@ -187,7 +187,7 @@ test('skip optional dependency that does not support the current pnpm version', 
     reason: 'unsupported_engine',
   })
   const reportedTimes = reporter.withArgs(logMatcher).callCount
-  expect(reportedTimes).toBe(1)
+  expect(reportedTimes).toBe(0)
 })
 
 test('don\'t skip optional dependency that does not support the current OS when forcing', async () => {

--- a/pkg-manager/headless/src/lockfileToHoistedDepGraph.ts
+++ b/pkg-manager/headless/src/lockfileToHoistedDepGraph.ts
@@ -185,7 +185,6 @@ async function fetchDeps (
         lockfileDir: opts.lockfileDir,
         nodeVersion: opts.nodeVersion,
         optional: pkgSnapshot.optional === true,
-        pnpmVersion: opts.pnpmVersion,
         supportedArchitectures: opts.supportedArchitectures,
       }) === false
     ) {

--- a/pkg-manager/package-requester/src/packageRequester.ts
+++ b/pkg-manager/package-requester/src/packageRequester.ts
@@ -235,7 +235,6 @@ async function resolveAndFetch (
             lockfileDir: options.lockfileDir,
             nodeVersion: ctx.nodeVersion,
             optional: wantedDependency.optional === true,
-            pnpmVersion: ctx.pnpmVersion,
             supportedArchitectures: options.supportedArchitectures,
           })
       )

